coderd/httpapi/httpapi.go

```diff
@@ -132,3 +135,81 @@ func WebsocketCloseSprintf(format string, vars ...any) string {
 
 	return msg
 }
+
+func SetupSSE(rw http.ResponseWriter, r *http.Request) error {
+	h := rw.Header()
+	h.Set("Content-Type", "text/event-stream")
+	h.Set("Cache-Control", "no-cache")
+	h.Set("Connection", "keep-alive")
+	h.Set("X-Accel-Buffering", "no")
+
+	f, ok := rw.(http.Flusher)
+	if !ok {
+		return xerrors.New("http.ResponseWriter is not http.Flusher")
+	}
+
+	_, err := io.WriteString(rw, ": ping\n\n")
+	if err != nil {
+		return err
+	}
+	f.Flush()
+
+	// Send a heartbeat every 15 seconds to avoid the connection being killed.
+	go func() {
+		ticker := time.NewTicker(time.Second * 15)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-r.Context().Done():
+				return
+			case <-ticker.C:
+				_, err := io.WriteString(rw, ": ping\n\n")
+				if err != nil {
+					return
+				}
+				f.Flush()
+			}
+		}
+	}()
```

Comment:
What's the behavior if this wrote at the same time as data was being written to the socket? Would you get a corrupted event?

Maybe instead of having callers call `Event()` to write events, you should just return an event writing function from this function which can acquire a lock throughout the entire write.